custom/icds_reports/sqldata/exports/poshan_progress_report.py

```diff
@@ -0,0 +1,267 @@
+            self.row_constants[4].remove('district_name')
+
+    def _generate_quarter_years(self):
+        months = []
+        end_month = self.quarter * 3
+        for i in range(end_month - 2, end_month + 1):
+            months.append(date(self.year, i, 1))
+        return months
+
+    def __handle_average(self, number):
+        if number is None:
+            ret = 0
+        else:
+            ret = "%.2f" % (number / 3)
+        return float(ret)
+
+    def __calculate_percent(self, num, den, extra_number):
+        if den == 0:
+            ret = 0
+        else:
+            ret = (num / den) * 100
+
+        if extra_number:
+            ret = ret / extra_number
+        return "{}%".format("%.2f" % ret)
+
+    def __calculate_percentage_in_rows(self, row, all_cols):
+        for k, v in COLS_PERCENTAGE_RELATIONS.items():
+            num = row[all_cols.index(v[0])]
+            den = row[all_cols.index(v[1])]
+            extra_number = v[2] if len(v) > 2 else None
+            row[all_cols.index(k)] = self.__calculate_percent(num, den, extra_number)
+        return row
+
+    def quarter_wise(self, filters, order_by, aggregation_level):
+        """
+        :param filters: quaterwise filter (months in [month array])
+        :param order_by: order by columns
+        :return: excel_rows
+        """
+        headers_comprehensive, cols_comprehensive, headers_summary, cols_summary, cols_to_fetch = self.row_constants
+        query_set = PoshanProgressReportView.objects.filter(**filters).order_by(*order_by)
+        all_cols = list(set(cols_comprehensive + cols_to_fetch))
+
+        # it used to uniquely identify the row
+        # for district wise data we need district id
+        # for statewise we need state id
+        unique_id = 'district_id'
+        if aggregation_level == 1:
+            unique_id = 'state_id'
+        cols = cols_to_fetch[:]
+        cols.append(unique_id)  # used as key for the dict
+        data = query_set.values(*cols)
+        row_data_dict = {}
+        dummy_row = [0 for _ in range(0, len(all_cols))]
+        headers = headers_comprehensive
+        # update the dict
+        # {'unique_id': [contains the excel row with sum of col values for all months eg. m1+m2+m3]}
+        for row in data:
+            if row[unique_id] not in row_data_dict.keys():
+                row_data_dict[row[unique_id]] = dummy_row[:]
+            row_data = row_data_dict[row[unique_id]][:]
+            for k, v in row.items():
+                if k in ['state_name', 'district_name']:
+                    row_data[all_cols.index(k)] = v
+                elif k != unique_id:
+                    row_data[all_cols.index(k)] += v if v else 0
+            row_data_dict[row[unique_id]] = row_data
+
+        # calculating average
+        # m1+m2+m3/3
+        total_row = [0 for _ in range(0, len(all_cols))]
+        for k, v in row_data_dict.items():
+            for col in all_cols[:]:
```

Comment:
you can use enumerate here so that do don't need to to find index inside loop explicitly. 
Also, Why are we creating a new array here using `[:]`

Reply:
updated